### DISCUSSION TITLE
Write out all parsed items into a single file instead of multiple.

### DISCUSSION THREAD

**Discussion by bigcurl** [bigcurl/ffi-cuda-toolkit-gem-generator#4] [General]
Hi @s96Abrar,

from my current understanding of how FFI it seems that all functions, structs, typedefs, enums, etc. have to be in a single file. While working on the enums I noticed that it is not possible to define an enum in one file and somehow include them into another. I think for now the best solution is to parse everything and than write it out to disk into one file (wrapper.rb), using one template.

Lets see how this approach works but just that you know that there are some fundamental changes to the code coming.

**Comment by s96Abrar** [comment]
Yeah I just noticed that CUDA driver API interface is written in one file named `cuda.h`

So yes, we need to write out all the functions, enums, struct in one file.

**Reply by bigcurl** [reply]
I think it is more a ruby loading problem than a C problem but I think you are right, all functions are in one namespace, so lets go with one namespace for now.

I think the namespace suggested in the docu is not bad, how they structured the functions and this was my first intention to have the same in the ruby code.

**Comment by bigcurl** [comment]
The code to parse everything into a single file is in the repo. The tests (3) should work now. You can run the tests by calling rake in the gem/cuda folder.

**Comment by s96Abrar** [comment]
Found something weird for a union structure in the documentation [CUstreamBatchMemOpParams_v1](https://docs.nvidia.com/cuda/archive/11.4.2/cuda-driver-api/unionCUstreamBatchMemOpParams__v1.html#unionCUstreamBatchMemOpParams__v1)

there is nothing defined in the doc for this union structure but I found some implementation code for that in `cuda.h`

```
/**
 * Per-operation parameters for ::cuStreamBatchMemOp
 */
typedef union CUstreamBatchMemOpParams_union {
    CUstreamBatchMemOpType operation;
    struct CUstreamMemOpWaitValueParams_st {
        CUstreamBatchMemOpType operation;
        CUdeviceptr address;
        union {
            cuuint32_t value;
            cuuint64_t value64;
        };
        unsigned int flags;
        CUdeviceptr alias; /**< For driver internal use. Initial value is unimportant. */
    } waitValue;
    struct CUstreamMemOpWriteValueParams_st {
        CUstreamBatchMemOpType operation;
        CUdeviceptr address;
        union {
            cuuint32_t value;
            cuuint64_t value64;
        };
        unsigned int flags;
        CUdeviceptr alias; /**< For driver internal use. Initial value is unimportant. */
    } writeValue;
    struct CUstreamMemOpFlushRemoteWritesParams_st {
        CUstreamBatchMemOpType operation;
        unsigned int flags;
    } flushRemoteWrites;
    cuuint64_t pad[6];
} CUstreamBatchMemOpParams_v1;
```

Confused what to do for this union? Should we define it manually?

**Comment by bigcurl** [comment]
I had the same for some items. I put it in manually. And there are some
errors in the documentation as well. That is why I think going with castxml
and parsing the header and xml is the better way to go.

Abrar ***@***.***> schrieb am Do., 21. Okt. 2021, 17:45:

**Comment by s96Abrar** [comment]
I think auto creation of the ruby interface (wrapper.rb) for CUDA driver API won't be a great thing. While parsing structs I saw the documentation did not pointed some of the union inside a struct. Although it is very efficient to create the wrapper easily.

How about writing only the struct manually?

I think my code for generating the struct are not quite good. If you have time can you give it a look?

**Reply by bigcurl** [reply]
It seems structs and unions are not so easy to parse from the documentation. I have created a new branch "castxml" to explorer the possibility to parse from an xml file. castxml uses LLVM to create an xml tree from the code. I added the xml files instead of the html docus. So far I have functions, typedefs and enums working. I think structs and unions will be far easier to parse from the xml. Also all elements that the struct is composed of are in the xml file.

Give it a try right, currently writing the wrapper out to file does not work but I will continue on Monday. It is just a matter of adding a few missing items.

**Comment by s96Abrar** [comment]
So far xml parsing the `cuda.h.xml` has lot of things than documentation. But I think we are not following one thing that C or C++ programming needed mostly. Parsing all the types such as typedef, #define 's, functions, structs separately will end up causing not finding types.

I tried to manually fix the documentation parsing and tried run the test but I end up having type error in ffi (the main branch). The error is all about we are placing those functions, structs, typedefs as where we wanted to. But this shouldn't be correct (I guess you understood).

Like if we use a new member (struct) within a struct and the new member is not defined before the struct then we end up having type error.

```
class tempStruct < FFI::Struct
  layout :value, NewStructType #<- We will get type error here
end
```

So I think we should use the namespace members attribute to parse each member sequentially.

**Reply by bigcurl** [reply]
Good point, I had not thought of the order issue. Let me finish the struct and function type parsing code, then next thing is the order.

**Comment by bigcurl** [comment]
Hi @s96Abrar ,

I'm stuck with the structs parsing code. 

Please take over. 

I uploaded my code that I have so far but given that we have to take the order in mind of the file, maybe you can separate the parsing out into different functions. When you parse the namespace members attribute keep in mind to skip those that are not in the cuda.h file. See the file attribute on each entry.

**Reply by s96Abrar** [reply]
OK. I will start doing it.

**Comment by bigcurl** [comment]
I found one more project today that more or less does what we want to achieve. 

`https://github.com/ffi/ffi-swig-generator`

It takes an xml file generated by `swig` and produces the ffi bindings. 
The `swig` output seems to have the advantage over the `castxml` output that it also includes the defines otherwise I would say they contain the same amount of information.

I would say lets use the new project and save some time BUT I think the output does not seem to work. Not sure if it is the generator that has not been updated in a while or it is the naming of the `cuda` lib.

You can checkout the output in file `gem/cuda/lib/cuda/11.5.0/driver_api.rb` in the repository. Otherwise you can generate the files yourself. 

Let me know what you think which approach looks best. Using the new swig output as input and write the generator code new or fix the existing one?

**Reply by s96Abrar** [reply]
Using it will be great as it is written and quite tested I guess. Fixing old one will be time consuming and we might miss some use cases (but cuda library interface is filled up with so many use cases to test out a C code parsing.)

But here is a problem I tried to run test but still there is some issues with `swig` generated ruby code. Some of the typedef of struct are not correct. I see that `swig` generated code is very outstanding but C code is very unexpected.

I still have in mind that we better write the ruby code manually auto generation of the code getting very complex.

**Reply by s96Abrar** [reply]
Another thing we can try, is we can fork the `swig` gem repo then try make the changes how to resolve those type errors.

**Comment by bigcurl** [comment]
I forked the repository (a newer fork of the project I mentioned earlier) here: https://github.com/bigcurl/ffi-swig-generator
We can apply our changes there. Not sure exactly what is wrong with the generated code but I think NVIDIA does something with the typedefs to apply internal code. This is why there are _v1 and _v2 spread throughout the code. It seems that some names are not correct in the header. Some structs have _St at the end but the typedef is missing. Maybe we need to add those by manually.

**Reply by s96Abrar** [reply]
Yeah there are so many wrong type names even generated code has a lot.

How can we use the forked repository to create the generated code?

**Comment by bigcurl** [comment]
Download `swig` http://www.swig.org/ . I used the Ubuntu Packaging System: `sudo apt install swig`
Run `swig`: `swig -xml -module cuda cuda.h`
This should output the xml file

git clone git@github.com:bigcurl/ffi-swig-generator.git
cd ffi-swig-generator
run `ruby bin/ffi-gen input output`

**Reply by s96Abrar** [reply]
OK Got it.

**Comment by s96Abrar** [comment]
It is not possible to change the `ffi-swig-generator` like i thought. Understanding the code will take so much time and swig generated xml file is also a bit complex to me.

What I think is we can generate the ruby code from the `ffi-swig-generator` then have to manually edit the struct. Because there is so many type errors for the struct and typedef of struct.

Also providing the auto-generation of the code won't be good idea as we have to manually edit the generated ruby code and make it working.

**Comment by bigcurl** [comment]
I think there is a patter with what is wrong with the structs. I noticed that the ones which are not working have St at the end and the typedef for mapping that to the name without St is missing. Either we provide those in a separate (manualy done) file or we use a find and replace command line tool and running over the output. But for now I would also go with a manual edited file. The tests are what is important to me since they show how to use the SDK. And of course we cannot start with the tests until the file runs in ruby. 

So better half automated but tested than automated and not tests.

**Reply by s96Abrar** [reply]
Well I see that is what we can do now half automated and half is manual.

Can you provide cuda.h for 11.5.0 version here?

**Reply by bigcurl** [reply]
Both the header and the xml output from swig are in the repo now

**Reply by s96Abrar** [reply]
OK Great.

**Reply by s96Abrar** [reply]
Where should I upload both edited driver_api.rb files?

**Reply by bigcurl** [reply]
Put it in gem/cuda/lib/cuda/11.5.0/driver_api.rb

**Reply by bigcurl** [reply]
Then the test should pick it up automaticaly

**Reply by s96Abrar** [reply]
Sorry, I mean in which branch? castxml or main branch?

**Reply by bigcurl** [reply]
Take the castxml branch. Wrong name but this is not so important.

**Reply by s96Abrar** [reply]
Try check the test

**Reply by bigcurl** [reply]
Great, the tests run. You can run the test by invoking `rake` in the `/gem/cuda` folder.
Can you mark your changes in the file? Maybe we spot a pattern or it can be resolved by including another file with the changes you made.

Other than that you can start working on the tests.

**Reply by s96Abrar** [reply]
My changes only contains adding extra `typedef` for the struct and there is some struct and union which are not in order.

Yes I tried running `rake` in the `gem/cuda/` folder and it says 2 failures.

What to do on the tests?

**Reply by bigcurl** [reply]
I updated the tests in the meantime, the new parser outputs the enums differently than they are in the header file.
CUDA_SUCCESS becomes success without the CUDA prefix.
 
> What to do on the tests?

If added already two test files which show how the syntax works. Basically it is writing one test for each function in the toolkit to show that it returns the expected result and shows how to use the Cuda Toolkit wrapper.

**Reply by s96Abrar** [reply]
So just getting the return code and matching them?

But I guess we can't do that for void returned functions.

**Reply by bigcurl** [reply]
Return code will always be :success  but as you can see in the test test_cu_device_get_count which checks the numbers of cuda devices in the system, a variable count_pointer has to be given which has a value after the call. I think there are no functions which return void in Cuda.
